### PR TITLE
[Autotuner] CI Smoke Test - Tune & Sweep

### PR DESCRIPTION
**Pull request by luarss**
@vvbandeira 
Goal is to do smoke test using CLI function call. 
Supports all platforms: asap7, sky130hd, ihp-sg13g2, gf180

- [x] Tune 
- [x] Sweep
- [x] Change Jenkinsfile and test_helper.sh

### CODE REVIEW

File: tools/AutoTuner/.gitignore
Comment (rovinski) [inline tools/AutoTuner/.gitignore:1]: How much of this is actually needed?

File: tools/AutoTuner/src/autotuner/__init__.py
Comment (rovinski) [inline tools/AutoTuner/src/autotuner/__init__.py:1]: empty file?

File: tools/AutoTuner/README.md
Comment (rovinski) [inline tools/AutoTuner/README.md:1]: Why link?

File: tools/AutoTuner/src/autotuner/__init__.py
Comment (luarss) [reply tools/AutoTuner/src/autotuner/__init__.py:1]: I included it because there was a discussion whether to make it a proper python package. Will be removing it for now.

File: tools/AutoTuner/.gitignore
Comment (luarss) [reply tools/AutoTuner/.gitignore:1]: It is a generic Python .gitignore, helped to catch some stray files while I was generating the autotuner python package. Will probably remove it.

File: tools/AutoTuner/README.md
Comment (luarss) [reply tools/AutoTuner/README.md:1]: Better visibility? The old documentation points to the the remote setup of Autotuner.

File: tools/AutoTuner/src/autotuner/distributed.py [hunk 1]
```diff
@@ -960,5 +961,8 @@ def sweep():
         if analysis.best_result['minimum'] == ERROR_METRIC:
             print('[ERROR TUN-0016] No successful runs found.')
             sys.exit(1)
+
+        # if successful
+        sys.exit(0)
```
Comment (vvbandeira) [inline tools/AutoTuner/src/autotuner/distributed.py]: Is this required?

File: tools/AutoTuner/test/smoke_test_sweep.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+# Disable GF180 tests for now.
+# class GF180SweepSmokeTest(BaseSweepSmokeTest):
+#     platform = "gf180"
+#     design = "ibex"
+
+#     def test_sweep(self):
+#         out = subprocess.run(self.command, shell=True, check=True)
+#         successful = out.returncode == 0
+#         self.assertTrue(successful)
```
Comment (vvbandeira) [inline tools/AutoTuner/test/smoke_test_sweep.py]: Remove code instead of commenting it out.

File: tools/AutoTuner/test/smoke_test_tune.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+# Disable GF180 tests for now.
+# class GF180TuneSmokeTest(BaseTuneSmokeTest):
+#     platform = "gf180"
+#     design = "ibex"
+
+#     def test_tune(self):
+#         out = subprocess.run(self.command, shell=True, check=True)
+#         successful = out.returncode == 0
+#         self.assertTrue(successful)
```
Comment (vvbandeira) [inline tools/AutoTuner/test/smoke_test_tune.py]: Remove code instead of commenting it out.

File: docs/user/InstructionsForAutoTuner.md [hunk 4]
```diff
@@ -193,6 +193,14 @@ We show three different views possible at the end, namely: `Table View`, `Scatte
...
 <p style="text-align: center;">Parallel Coordinate View (best run is in green)</p>
 
+## Testing framework
+
+```
+# Smoke test
+cd tools/AutoTuner/tests
+python3 smoke_test_tune.py
+python3 smoke_test_sweep.py
+```
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]: Shouldn't there be instructions on the virtual env setup?

**Review by vvbandeira — CHANGES_REQUESTED** [review]

File: tools/AutoTuner/src/autotuner/distributed.py (hunk 1 above)
Comment (luarss) [inline tools/AutoTuner/src/autotuner/distributed.py]:
```suggestion
```

File: tools/AutoTuner/src/autotuner/distributed.py [hunk 5]
```diff
@@ -882,6 +882,7 @@ def sweep():
     print('[INFO TUN-0009] Waiting for results.')
     ray.get(workers)
     print('[INFO TUN-0010] Sweep complete.')
+    sys.exit(0)
```
Comment (vvbandeira) [inline tools/AutoTuner/src/autotuner/distributed.py]: This should not be needed as the script will end after the call of `sweep()`. We should only exit with an error in places other than `__main__.`

File: docs/user/InstructionsForAutoTuner.md [hunk 6]
```diff
@@ -30,7 +30,7 @@ which is a free minimal installer for the package manager `conda`.
 
 ```shell
 # set up conda environment
-conda create -n autotuner_env python=3.9
+conda create -n autotuner_env python=3.8 -y
 conda activate autotuner_env
 
 # install requirements
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]: A simple `setup.sh` would be helpful here. Users can use a single command to set up, and CI can use the same.

File: docs/user/InstructionsForAutoTuner.md [hunk 7]
```diff
@@ -193,6 +193,17 @@ We show three different views possible at the end, namely: `Table View`, `Scatte
...
+## Testing framework
+
+```
+# Activate your virtualenv before this
+conda activate autotuner_env
+
+# Smoke test
+cd tools/AutoTuner/tests
+python3 smoke_test_tune.py
+python3 smoke_test_sweep.py
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]: Could the tests themselves call the `setup.sh` (as suggested above) to activate the environment?

Comment (luarss) [reply docs/user/InstructionsForAutoTuner.md]: They can, but from the same principles of ORFS can we leave it optional? Similar to how we have `Env.sh` as optional

File: docs/user/InstructionsForAutoTuner.md (hunk 6 above)
Comment (luarss) [reply docs/user/InstructionsForAutoTuner.md]: `. autotuner_env/bin/activate`

File: docs/user/InstructionsForAutoTuner.md (hunk 7 above)
Comment (luarss) [reply docs/user/InstructionsForAutoTuner.md]: Update the docs here. 3 step instructions
- install depen
- source env
- run test.

File: flow/test/test_helper.sh [hunk 8]
```diff
@@ -63,4 +63,36 @@ if [ ! -z ${MAKE_ISSUE+x} ]; then
...
+if [ $RUN_AUTOTUNER -eq 1 ]; then
+  case $PLATFORM in
+       "gf180" | "nangate45" | "sky130hd_fakestack" | "sky130hs")
+        RUN_AUTOTUNER=0
+        ;;
+  esac
+fi
+
+if [ $RUN_AUTOTUNER -eq 1 ]; then
+  # change directory to ../
```
Comment (vvbandeira) [inline flow/test/test_helper.sh]:
```suggestion
  # change directory to the root of the repo
```

File: flow/test/test_helper.sh [hunk 9]
```diff
@@ -63,4 +63,36 @@ if [ ! -z ${MAKE_ISSUE+x} ]; then
...
+
+  PLATFORM=${PLATFORM//-/}
+  PLATFORM=${PLATFORM^^}  
+
+  echo "Running Autotuner smoke tune test"
+  python3 -m unittest tools.AutoTuner.test.smoke_test_tune.${PLATFORM^^}TuneSmokeTest.test_tune
+  
+  echo "Running Autotuner smoke sweep test"
+  python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM^^}SweepSmokeTest.test_sweep
+fi
```
Comment (vvbandeira) [inline flow/test/test_helper.sh]:
``` suggestion
if [ $RUN_AUTOTUNER -eq 1 ]; then
  # change directory to the root of the repo
  cd ..
  echo "Install dependencies in Venv"
  ./tools/AutoTuner/installer.sh
  ./tools/AutoTuner/setup.sh

  # remove dashes
  PLATFORM=${PLATFORM//-/}
  # convert to uppercase
  PLATFORM=${PLATFORM^^}

  echo "Running Autotuner smoke tune test"
  python3 -m unittest tools.AutoTuner.test.smoke_test_tune.${PLATFORM}TuneSmokeTest.test_tune

  echo "Running Autotuner smoke sweep test"
  python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM}SweepSmokeTest.test_sweep
fi
```

File: tools/AutoTuner/setup.sh [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+# Get the directory of the current script
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```
Comment (vvbandeira) [inline tools/AutoTuner/setup.sh]:
```suggestion
SCRIPT_DIR="$(dirname "${BASH_SOURCE[0]}")"
```

**Review by vvbandeira — APPROVED** [review]

**Comment by vvbandeira** [comment]
@luarss 
CI is failing to find AT files:
```
Install dependencies in Venv
+ ./tools/AutoTuner/installer.sh
flow/test/test_helper.sh: line 85: ./tools/AutoTuner/installer.sh: No such file or directory
```

**Comment by luarss** [comment]
> @luarss CI is failing to find AT files:
> 
> ```
> Install dependencies in Venv
> + ./tools/AutoTuner/installer.sh
> flow/test/test_helper.sh: line 85: ./tools/AutoTuner/installer.sh: No such file or directory
> ```

This is because the installer script is in the other PR #2025

**Comment by vvbandeira** [comment]
@luarss 
Please fix merge conflict.

**Comment by luarss** [comment]
@vvbandeira I see that `./tools/AutoTuner` files are not being copied into the docker image, see the latest run: https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Public/job/PR-2042-merge/74/pipeline-console/?selected-node=164

Any ideas?

File: tools/AutoTuner/.gitignore
Comment (vvbandeira) [inline tools/AutoTuner/.gitignore:1]: Please remove comments and clean up the list to only include what is required for our use case.

File: tools/AutoTuner/test/smoke_test_sweep.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        self.command = (
+            "python3 distributed.py"
+            f" --design {self.design}"
+            f" --platform {self.platform}"
+            f" --experiment {self.experiment}"
+            f" --config {self.config}"
+            f" --jobs {self.jobs}"
+            f" sweep"
+        )
+
```
Comment (vvbandeira) [inline tools/AutoTuner/test/smoke_test_sweep.py:42]: Add the `test_sweep` method and throw an exception in the body saying that this needs to be implemented in the derivative classes.

File: tools/AutoTuner/test/smoke_test_tune.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        self.experiment = f"smoke-test-tune-{self.platform}"
+        self.command = (
+            "python3 distributed.py"
+            f" --design {self.design}"
+            f" --platform {self.platform}"
+            f" --experiment {self.experiment}"
+            f" --config {self.config}"
+            f" tune --samples 1"
+        )
+
```
Comment (vvbandeira) [inline tools/AutoTuner/test/smoke_test_tune.py:28]: Same as above.

File: tools/AutoTuner/.gitignore [hunk 13]
```diff
@@ -1,2 +1,164 @@
...
+cython_debug/
+
+# PyCharm
+#  JetBrains specific template is maintained in a separate JetBrains.gitignore that can
+#  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
+#  and can be added to the global gitignore or merged into this file.  For a more nuclear
+#  option (not recommended) you can uncomment the following to ignore the entire idea folder.
+#.idea/
+
+>>>>>>> e0148686 (working smoke test for tune)
```
Comment (vvbandeira) [inline tools/AutoTuner/.gitignore]: Remove?

**Review by vvbandeira — CHANGES_REQUESTED** [review]

**Comment by luarss** [comment]
Seems like the ./tools/AutoTuner files are still not copied? https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Public/job/PR-2042-head/15/pipeline-console/?selected-node=164

File: docker/Dockerfile.builder [hunk 14]
```diff
@@ -33,6 +33,7 @@ COPY --link flow/platforms flow/platforms
 COPY --link flow/util flow/util
 COPY --link flow/scripts flow/scripts
 COPY --link flow/designs flow/designs
+COPY -- link tools/AutoTuner tools/AutoTuner
```
Comment (vvbandeira) [inline docker/Dockerfile.builder]:
```suggestion
COPY --link tools/AutoTuner tools/AutoTuner
```

**Review by vvbandeira — CHANGES_REQUESTED** [review]
Looks like you had a typo

File: tools/AutoTuner/installer.sh
Comment (vvbandeira) [inline tools/AutoTuner/installer.sh:1]: Why do we need this empty file?

Comment (luarss) [reply tools/AutoTuner/installer.sh:1]: Not an empty file, I was just making it executable (chmod +x)

File: docker/Dockerfile.builder [hunk 15]
```diff
@@ -33,10 +33,12 @@ COPY --link flow/platforms flow/platforms
 COPY --link flow/util flow/util
 COPY --link flow/scripts flow/scripts
 COPY --link flow/designs flow/designs
+COPY --link tools/AutoTuner tools/AutoTuner
 
 COPY --link --from=orfs-builder-base /OpenROAD-flow-scripts/tools/install tools/install
 COPY --link \
-     --exclude=.git* --exclude=tools/ --exclude=docs/ --exclude=docker/ \
+     --exclude=.git* --exclude=tools/OpenROAD --exclude=tools/codespace \
+     --exclude=tools/yosys --exclude=tools/yosys_utils --exclude=docs/ --exclude=docker/ \
```
Comment (vvbandeira) [inline docker/Dockerfile.builder]: I think we can revert this change, AT is already copied in line 36.

File: tools/AutoTuner/installer.sh [hunk 16]
```diff
@@ -6,6 +6,7 @@ cd $script_dir
 
 # Define the virtual environment name
 venv_name="autotuner_env"
+required_version="3.8"
```
Comment (vvbandeira) [inline tools/AutoTuner/installer.sh]: Didn't we agree somewhere to drop the specific version and keep with either `python` or `python3`?

Comment (luarss) [reply tools/AutoTuner/installer.sh]: `python` did not resolve to any command in previous iteration, maybe I will do 

```
required_version="3"
python${required_version} ...
```

File: tools/AutoTuner/installer.sh [hunk 17]
```diff
@@ -2,11 +2,11 @@
 
 # Get the directory where the script is located
 script_dir="$(dirname "${BASH_SOURCE[0]}")"
-cd $script_dir
 
 # Define the virtual environment name
 venv_name="autotuner_env"
+required_version="3"
 python${required_version} -m venv "$script_dir/$venv_name"
```
Comment (vvbandeira) [inline tools/AutoTuner/installer.sh]:
```suggestion
python3 -m venv "$script_dir/$venv_name"
```

File: flow/test/test_helper.sh [hunk 18]
```diff
@@ -77,4 +77,38 @@ if [ $ret -eq 0 ] && grep -q 'power:' <(echo $TARGETS); then
...
+    RUN_AUTOTUNER=1
+    ;;
+esac
+if [ $RUN_AUTOTUNER -eq 1 ]; then
+  case $PLATFORM in
+       "gf180" | "nangate45" | "sky130hd_fakestack" | "sky130hs")
+        RUN_AUTOTUNER=0
+        ;;
+  esac
+fi
```
Comment (vvbandeira) [inline flow/test/test_helper.sh]:
```suggestion
# Run Autotuner CI specifically for gcd on selected platforms.
RUN_AUTOTUNER=0
case $DESIGN_NAME in
  "gcd")
    RUN_AUTOTUNER=1
    ;;
esac
case $PLATFORM in
     "asap7" | "sky130hd" | "ihp-sg13g2" )
      # Keep RUN_AUTOTUNER enabled only for these platforms
      ;;
     *)
      RUN_AUTOTUNER=0
      ;;
esac
```

**Review by vvbandeira — CHANGES_REQUESTED** [review]

**Review by vvbandeira — APPROVED** [review]
